internal/config/rule.go

```diff
@@ -1,195 +1,112 @@
 package config
 
 import (
-	"context"
-	"github.com/icinga/icinga-notifications/internal/filter"
+	"fmt"
 	"github.com/icinga/icinga-notifications/internal/rule"
-	"github.com/jmoiron/sqlx"
-	"go.uber.org/zap"
 	"slices"
 )
 
-func (r *RuntimeConfig) fetchRules(ctx context.Context, tx *sqlx.Tx) error {
-	var rulePtr *rule.Rule
-	stmt := r.db.BuildSelectStmt(rulePtr, rulePtr)
-	r.logger.Debugf("Executing query %q", stmt)
-
-	var rules []*rule.Rule
-	if err := tx.SelectContext(ctx, &rules, stmt); err != nil {
-		r.logger.Errorln(err)
-		return err
-	}
-
-	rulesByID := make(map[int64]*rule.Rule)
-	for _, ru := range rules {
-		ruleLogger := r.logger.With(zap.Inline(ru))
+func (r *RuntimeConfig) applyPendingRules() {
+	incrementalApplyPending(
+		r,
+		&r.Rules, &r.configChange.Rules,
+		func(newElement *rule.Rule) error {
+			if newElement.TimePeriodID.Valid {
+				timePeriod, ok := r.TimePeriods[newElement.TimePeriodID.Int64]
+				if !ok {
+					return fmt.Errorf("rule refers unknown time period %d", newElement.TimePeriodID.Int64)
+				}
 
-		if ru.ObjectFilterExpr.Valid {
-			f, err := filter.Parse(ru.ObjectFilterExpr.String)
-			if err != nil {
-				ruleLogger.Warnw("ignoring rule as parsing object_filter failed", zap.Error(err))
-				continue
+				newElement.TimePeriod = timePeriod
 			}
 
-			ru.ObjectFilter = f
-		}
-
-		ru.Escalations = make(map[int64]*rule.Escalation)
-
-		rulesByID[ru.ID] = ru
-		ruleLogger.Debugw("loaded rule config")
-	}
-
-	var escalationPtr *rule.Escalation
-	stmt = r.db.BuildSelectStmt(escalationPtr, escalationPtr)
-	r.logger.Debugf("Executing query %q", stmt)
-
-	var escalations []*rule.Escalation
-	if err := tx.SelectContext(ctx, &escalations, stmt); err != nil {
-		r.logger.Errorln(err)
-		return err
-	}
-
-	escalationsByID := make(map[int64]*rule.Escalation)
-	for _, escalation := range escalations {
-		escalationLogger := r.logger.With(zap.Inline(escalation))
-
-		rule := rulesByID[escalation.RuleID]
-		if rule == nil {
-			escalationLogger.Warnw("ignoring escalation for unknown rule_id")
-			continue
-		}
-
-		if escalation.ConditionExpr.Valid {
-			cond, err := filter.Parse(escalation.ConditionExpr.String)
-			if err != nil {
-				escalationLogger.Warnw("ignoring escalation, failed to parse condition", zap.Error(err))
-				continue
+			newElement.Escalations = make(map[int64]*rule.Escalation)
+			return nil
+		},
+		func(curElement, update *rule.Rule) error {
+			curElement.IsActive = update.IsActive
+			curElement.Name = update.Name
+			curElement.ObjectFilter = update.ObjectFilter
```

Comment:
Can't `rule#TimePeriodID` also be changed at runtime?

Reply:
Good catch!